### PR TITLE
Context overhaul

### PR DESCRIPTION
**Pull request by david-andrew**
- Merged context creation functions `add_permanent_context`, `add_timed_context`, and `add_managed_context` all into a single `add_context` function that returns the id of the context message made

  ```python
  def add_context(self, message, lifetime=-1) -> int
      ...
  ```
- Added `clear_context` function for deleting context messages given their id (this replaces the functionality of the `add_managed_context` deleter callback)
- Made contexts internally maintain their id + lifetime for cleaner management
- renamed the states in the `LoopController` class to be more clear

@mattprintz I tested only a little bit. Everything looks good, but when you get a chance, let me know if this works for you.

### CODE REVIEW

File: archytas/agent.py [hunk 1]
```diff
@@ -65,86 +72,64 @@ def __init__(self, *, model:str='gpt-4', prompt:str="You are a helpful assistant
...
-        context_message = {"role": Role.system, "content": context}
-        self.messages.append(context_message)
-        self._all_context_messages.append(context_message)
+        #filter out messages that have expired
+        self.messages = [message for message in self.messages if not isinstance(message, ContextMessage) or message.lifetime != 0]
 
+        #decrement lifetimes of all timed context messages
+        for message in self.messages:
+            if isinstance(message, ContextMessage):
+                message.lifetime -= 1
```
Comment (mattprintz) [inline archytas/agent.py:110]: This is also decrementing values that have defaulted to `-1`.

File: archytas/agent.py [hunk 2]
```diff
@@ -65,86 +72,64 @@ def __init__(self, *, model:str='gpt-4', prompt:str="You are a helpful assistant
...
-        context_message = {"role": Role.system, "content": context}
-        self.messages.append(context_message)
-        self._all_context_messages.append(context_message)
-
-        def remove_context():
-            self.messages.remove(context_message)
-            self._all_context_messages.remove(context_message)
-
-        return remove_context
+        self.messages = [message for message in self.messages if not isinstance(message, ContextMessage) or message.id != id]
```
Comment (mattprintz) [inline archytas/agent.py:119]: Replace confusing logic here too, please.

File: archytas/agent.py (hunk 1 above)
Comment (david-andrew) [reply archytas/agent.py:110]: it's true it decrements values that are `-1` but it doesn't matter. I guess if you want it to be explicit, I could have lifetime be `int|None`

**Review by mattprintz — APPROVED** [review]
:+1:

**Pull request merged**